File: ypsopump/src/main/java/info/nightscout/androidaps/plugins/pump/ypsopump/driver/YpsopumpPumpStatus.java
package info.nightscout.androidaps.plugins.pump.ypsopump.driver;

import org.jetbrains.annotations.NotNull;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

import javax.inject.Inject;
import javax.inject.Singleton;

import info.nightscout.androidaps.plugins.bus.RxBusWrapper;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpDeviceState;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType;
import info.nightscout.androidaps.plugins.pump.ypsopump.defs.YpsoPumpFirmware;
import info.nightscout.androidaps.plugins.pump.ypsopump.util.YpsoPumpConst;
import info.nightscout.androidaps.utils.resources.ResourceHelper;
import info.nightscout.androidaps.utils.sharedPreferences.SP;


/**
 * Created by andy on 1/4/21.
 */

@Singleton
public class YpsopumpPumpStatus extends info.nightscout.androidaps.plugins.pump.common.data.PumpStatus  {

    private final ResourceHelper resourceHelper;
    private final SP sp;
    private final RxBusWrapper rxBus;

    public String errorDescription = null;
    public YpsoPumpFirmware ypsopumpFirmware;
    public double baseBasalRate = 0.0d;


    public String serialNumber;
    public String pumpFrequency = null;
    public Double maxBolus;
    public Double maxBasal;

    // statuses
    private PumpDeviceState pumpDeviceState = PumpDeviceState.NeverContacted;
    public Date tempBasalStart;
    public Double tempBasalAmount = 0.0d;

    // fixme
    public Integer tempBasalLength = 0;

//    public BasalProfileStatus basalProfileStatus = BasalProfileStatus.NotInitialized;
//    public BatteryType batteryType = BatteryType.None;




    @Inject
    public YpsopumpPumpStatus(ResourceHelper resourceHelper,
                              SP sp,
                              RxBusWrapper rxBus
    ) {
        super(PumpType.YpsoPump);
        this.resourceHelper = resourceHelper;
        this.sp = sp;
        this.rxBus = rxBus;
        initSettings();
    }


    public void initSettings() {
        this.activeProfileName = "A";
        this.reservoirRemainingUnits = 75d;
        this.reservoirFullUnits = 160;
        this.batteryRemaining = 75;
        this.lastConnection = sp.getLong(YpsoPumpConst.Statistics.LastGoodPumpCommunicationTime, 0L);
        this.lastDataTime = this.lastConnection;
    }


    public double getBasalProfileForHour() {
        if (basalsByHour != null) {
            GregorianCalendar c = new GregorianCalendar();
            int hour = c.get(Calendar.HOUR_OF_DAY);

            return basalsByHour[hour];
        }

        return 0;
    }


    @NotNull
    public String getErrorInfo() {
        return (errorDescription == null) ? "-" : errorDescription;
    }


    public PumpDeviceState getPumpDeviceState() {
        return pumpDeviceState;
    }


    public void setPumpDeviceState(PumpDeviceState pumpDeviceState) {
        this.pumpDeviceState = pumpDeviceState;

//        rileyLinkUtil.getRileyLinkHistory().add(new RLHistoryItem(pumpDeviceState, RileyLinkTargetDevice.MedtronicPump));
//
//        rxBus.send(new EventRileyLinkDeviceStatusChange(pumpDeviceState));
    }
}

File: ypsopump/src/main/java/info/nightscout/androidaps/plugins/pump/ypsopump/driver/YpsopumpPumpStatus.kt
package info.nightscout.androidaps.plugins.pump.ypsopump.driver

import info.nightscout.androidaps.interfaces.Profile
import info.nightscout.androidaps.interfaces.PumpDescription
import info.nightscout.androidaps.plugins.bus.RxBus
import info.nightscout.androidaps.plugins.pump.common.data.PumpStatus
import info.nightscout.androidaps.plugins.pump.common.defs.BasalProfileStatus
import info.nightscout.androidaps.plugins.pump.common.defs.PumpDeviceState
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType
import info.nightscout.androidaps.plugins.pump.ypsopump.comm.data.YpsoPumpStatusEntry
import info.nightscout.androidaps.plugins.pump.ypsopump.comm.data.YpsoPumpStatusList
import info.nightscout.androidaps.plugins.pump.ypsopump.data.BasalProfileDto
import info.nightscout.androidaps.plugins.pump.ypsopump.defs.YpsoPumpFirmware
import info.nightscout.androidaps.plugins.pump.ypsopump.util.YpsoPumpConst
import info.nightscout.androidaps.utils.resources.ResourceHelper
import info.nightscout.shared.sharedPreferences.SP
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by andy on 1/4/21.
 */
@Singleton
class YpsopumpPumpStatus @Inject constructor(private val resourceHelper: ResourceHelper,
                                             private val sp: SP,
                                             private val rxBus: RxBus
) : PumpStatus(PumpType.YPSOPUMP) {

    lateinit var pumpDescription: PumpDescription
    var errorDescription: String? = null
    var ypsopumpFirmware: YpsoPumpFirmware = YpsoPumpFirmware.VERSION_1_5  // TODO we default to 1.0 here
    var isFirmwareSet = false
    @JvmField var baseBasalRate = 0.0
    var serialNumber: Long? = null
    var ypsoPumpStatusList: YpsoPumpStatusList? = null

    //        rileyLinkUtil.getRileyLinkHistory().add(new RLHistoryItem(pumpDeviceState, RileyLinkTargetDevice.MedtronicPump));
//
//        rxBus.send(new EventRileyLinkDeviceStatusChange(pumpDeviceState));
    //public Double maxBolus;
    //public Double maxBasal;
    // statuses
    var pumpDeviceState = PumpDeviceState.NeverContacted

    //@JvmField var tempBasalStart: Date? = null
    @JvmField var tempBasalPercent = 100
    @JvmField var tempBasalDuration = 0

    //var tempBasalEnd = 0L
    // TODO determine if this is needed
    var basalProfileStatus = BasalProfileStatus.NotInitialized
    var basalProfile: Profile? = null

    //var connectionStatus = YpsoConnectionStatus.NOT_CONNECTED
    var settingsServiceVersion: String? = null

    var lastConfigurationUpdate: Long = 0
    var configChanged: Boolean = false
    var bolusStep: Double = 0.1
    var maxBolus: Double? = null
    var maxBasal: Double? = null

    var forceRefreshBasalProfile: Boolean = true
    var basalProfilePump: BasalProfileDto? = null

    override fun initSettings() {
        activeProfileName = "A"
        reservoirRemainingUnits = 75.0
        reservoirFullUnits = 160
        batteryRemaining = 75
        lastConnection = sp.getLong(YpsoPumpConst.Statistics.LastGoodPumpCommunicationTime, 0L)
        lastDataTime = lastConnection
    }

    //var ypsoPumpStatusList: YpsoPumpStatusList? = null

    fun getPumpStatusValuesForSelectedPump(): YpsoPumpStatusEntry? {
        return ypsoPumpStatusList!!.map.get(serialNumber)
    }

    fun setPumpStatusValues(entry: YpsoPumpStatusEntry) {
        ypsoPumpStatusList!!.map.put(entry.serialNumber, entry)
        
    }

    val basalProfileForHour: Double
        get() {
            if (basalsByHour != null) {
                val c = GregorianCalendar()
                val hour = c[Calendar.HOUR_OF_DAY]
                return basalsByHour!![hour]
            }
            return 0.0
        }

    override val errorInfo: String
        get() = if (errorDescription == null) "-" else errorDescription!!

    init {
        initSettings()
    }
}